### ISSUE TITLE
Improve the definition of `evil-goto-definition`

### ISSUE THREAD

**Issue by nnicandro** [emacs-evil/evil#1103]
# Issue type

- Bug

# Environment

Emacs version: GNU Emacs 26.1 (build 1, x86_64-apple-darwin17.7.0, Carbon Version 158 AppKit 1561.6) of 2018-09-26
Operating System: macOS Sierra
Evil version: evil-git-6fde982
Evil installation type: El-Get
Graphical/Terminal: Graphical
Tested in a `make emacs` session (see [CONTRIBUTING.md](https://github.com/emacs-evil/evil/blob/master/CONTRIBUTING.md#bug-reports)): No

# Reproduction steps
- Start Emacs
- Call `evil-goto-definition` while `imenu--make-index-alist` is defined

# Expected behavior

If neither `imenu`, `semantic`, or `xref` find a definition then the fallback of searching the buffer should be called.

# Actual behavior

If `imenu--make-index-alist` is called, the fallback will never run. Also if `imenu` and `semantic` fail, `xref` will always be used without trying a "search the buffer" fallback if it fails.

# Further notes

`evil-goto-definition` probably needs to be re-written to something like

```emacs-lisp
(cond
  ((try imenu return t if success))
  ((try semantic return t if success))
  ((try xref return t if success))
  (t (search the buffer)))
```
There would still be an issue with handling the `etags--xref-backend` in `xref-backend-functions` since its behavior is to always ask for a TAG file if none could be found. I think that should be considered as a failure for xref and just try to search the buffer.

**Comment by redguardtoo** [comment]
I could contribute my `my-evil-goto-definition`.

`gd` did work differently with `vim` since some version. I agree the local buffer search should high the highest priority.

https://github.com/redguardtoo/emacs.d/blob/master/lisp/init-evil.el#L318

**Comment by ehahn** [comment]
`xref` shouldn't be called at all. `gd` only searches the current file ("If included files also need to be searched use the commands listed in include-search.")

**Comment by dzop** [comment]
Ah yes, looking at the Vim help it does mention that `gd` only searches the current buffer. I was mainly focused on how Evil was handling it. In that case we should probably remove the semantic and xref handling since both can jump to a different file.

We should also adhere to the Vim definition, specifically searching the local function first and if that fails, search the whole buffer.

**Comment by redguardtoo** [comment]
Ok, I will send a pull request. Sorry I totally forget this issue. Thanks for reminding me.

**Comment by redguardtoo** [comment]
Here is my pull request: https://github.com/emacs-evil/evil/pull/1161

**Comment by axelf4** [comment]
Fixed by #873 and follow-up PRs.
